Always go via the GOT for symbol references on AArch64

mshinwell opened the pull request:
This pull request should provide a first, rather naive solution for the problems in #1268.  It arranges for all symbol references, when `Config.dlcode` is set, to go via the GOT.  This passes the testsuite on an AArch64 machine running CentOS 7.3 but using binutils 2.29.

I ran into two problems as noted in the code.  These may be due to my own misunderstandings and the fact I haven't had time to read the ARM documentation in complete detail (about five hours spent on this pull request as it is):

1. If you try to reference a non-global symbol via the GOT, then what seems to happen is that a GOT entry is not allocated, and the relocation is evaluated to the GOT entry of some other random symbol.  This seems jolly unhelpful to me.  I would have liked to seen an error.

2. I tried using the GOT relocations in conjunction with "symbol + offset" (for the `Ibased` case) as I thought the ARM documentation suggested might work.  However this doesn't seem to work, with the offset being dropped.  I now add the offset manually.

In the medium term we should try to fix all this properly (although I still don't understand how the hidden symbol and alias solution works if the symbol's data is mutable).  However maybe this will do for now.

mshinwell committed:
AArch64 GOT fixed

xavierleroy commented on asmcomp/arm64/emit.mlp:
```diff
@@ -114,7 +114,8 @@ let emit_addressing addr r =
       `[{emit_reg r}, #{emit_int ofs}]`
   | Ibased(s, ofs) ->
-      `[{emit_reg r}, #:lo12:{emit_symbol_offset s ofs}]`
+      if !Clflags.dlcode then `[{emit_reg r}, #{emit_int ofs}]`
```
This will be rejected by the assembler if `ofs` is too large.  However, I think you can just `assert (not !Clflags.dlcode)` there, see my general comment on this PR.

xavierleroy commented on asmcomp/arm64/emit.mlp:
```diff
@@ -437,7 +438,11 @@ module BR = Branch_relaxation.Make (struct
+        match addr with
+        | Iindexed _ -> 0
+        | Ibased _ -> if !Clflags.dlcode then 2 else 1
+      in
```
Change not required if we ensure that `Ibased` mode is not used in `dlcode` mode.

xavierleroy commented on asmcomp/arm64/emit.mlp:
```diff
@@ -616,8 +621,17 @@ let emit_instr i =
+                `	adrp	{emit_reg reg_tmp1}, :got:{emit_symbol s}\n`;
+                `	ldr	{emit_reg reg_tmp1}, [{emit_reg reg_tmp1}, #:got_lo12:{emit_symbol s}]\n`
+              end;
+              reg_tmp1
```
Again, we could just have `assert (not !Clflags.dlcode)` here.

xavierleroy commented on asmcomp/arm64/emit.mlp:
```diff
@@ -643,8 +657,14 @@ let emit_instr i =
+                `	ldr	{emit_reg reg_tmp1}, [{emit_reg reg_tmp1}, #:got_lo12:{emit_symbol s}]\n`
+              end;
+              reg_tmp1
+	in
```
Likewise.  Perhaps this code could be factored between `Iload` and `Istore`, but we'll see later.

xavierleroy commented on asmcomp/arm64/emit.mlp:
```diff
@@ -931,7 +951,15 @@ let fundecl fundecl =
+         relocations evaluate to random other GOT entries.  For the moment
+         force all symbols to be global. *)
+      `	.globl	{emit_symbol s}\n`;
+    end;
```
This is unfortunate but so be it.

xavierleroy reviewed (commented):
I'm very happy to see this proposal, thanks a lot.

I think you missed a big simplification: `Ibased` addressing mode is not supposed to be used when we have to go through the GOT.  See file `asmcomp/arm64/selection.ml`, line 85, function `use_direct_addressing`.  You should remove the `|| Compilenv.symbol_in_current_unit symb` part of this function.  Then `Ibased` addressing will never be selected in PIC mode, and your changes to `emit.mlp` can be vastly simplified.

mshinwell commented:
Thanks for the review.  I've addressed all the comments; the testsuite still passes on the setup described above.  I have squashed the commits since it's easier just to read the whole thing again rather than the diff on top of the previous version.

Please merge if this is ok...

xavierleroy reviewed (approved):
Looks fine to me.  Thanks!

rwmjones commented:
I will give this a try tomorrow :-)

rwmjones commented:
What I did to test this (this is all on Fedora 28 on an aarch64 host):

 * Rebuilt the base Fedora ocaml compiler package substituting this patch instead of my own fix.
 * Rebuilt about a dozen of the largest/most complex Fedora OCaml packages, including Coq and Frama-C, using the ocaml compiler package built in the previous step.
 * Rebuilt Frama-C a second time using the the rebuilt Coq.

All of that worked with no problems.

I will go with this patch instead of my own for Fedora 27/28 and RHEL 7.5 since it seems to be a safer fix.
